File: problems/518_coin-change-ii/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int change(int amount, vector<int> &coins)
    {

        vector<long long> dp(amount + 1, 0);

        dp[0] = 1;

        for (int x : coins)
        {
            for (int i = x; i <= amount; i++)
            {
                if (dp[i] < INT_MAX)
                    dp[i] += dp[i - x];
            }
        }

        return dp[amount];
    }
};

int main()
{
    Solution solution;

    vector<pair<pair<int, vector<int>>, int>> samples = {
        {{5, {1, 2, 5}}, 4},
        {{3, {2}}, 0},
        {{10, {10}}, 1},
    };

    for (size_t i = 0; i < samples.size(); ++i)
    {
        int amount = samples[i].first.first;
        vector<int> coins = samples[i].first.second;
        int expected = samples[i].second;

        int result = solution.change(amount, coins);
        cout << "Example " << (i + 1) << " result: " << result
             << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/518_coin-change-ii/main.rs
struct Solution;

impl Solution {
    pub fn change(amount: i32, coins: Vec<i32>) -> i32 {
        let amount = amount as usize;
        let mut dp = vec![0i64; amount + 1];
        dp[0] = 1;
        for &c in &coins {
            let coin = c as usize;
            if coin == 0 || coin > amount { continue; }
            for j in coin..=amount {
                dp[j] += dp[j - coin];
            }
        }
        dp[amount].min(i32::MAX as i64) as i32
    }
}

fn main() {
    let samples: Vec<((i32, Vec<i32>), i32)> = vec![
        ((5, vec![1, 2, 5]), 4),
        ((3, vec![2]), 0),
        ((10, vec![10]), 1),
    ];

    for (idx, ((amount, coins), expected)) in samples.iter().enumerate() {
        let result = Solution::change(*amount, coins.clone());
        println!(
            "Example {} result: {}, expected: {}",
            idx + 1,
            result,
            expected
        );
    }
}
